core/src/main/java/com/backblaze/b2/json/B2JsonBoundedByteArrayOutputStream.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2021, Backblaze Inc. All Rights Reserved.
+ * License https://www.backblaze.com/using_b2_code.html
+ */
+package com.backblaze.b2.json;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UnsupportedEncodingException;
+import java.util.Arrays;
+
+/**
+ * A B2Json implementation of OutputStream that
+ * - stores stream content in a byte array
+ * - expands capacity when needed (typically doubles)
+ * - has a max capacity threshold
+ * - throws IOException if the threshold is reached
+ *
+ * THREAD-SAFE
+ */
+public class B2JsonBoundedByteArrayOutputStream extends OutputStream {
+    private static final int INITIAL_CAPACITY = 64;
+
+    // byte array to hold output content
+    private byte[] output;
+
+    // the current number of bytes written so far
+    private int size;
+
+    // maximum capacity the output array is allowed to grow to
+    private static final int MAX_CAPACITY = Integer.MAX_VALUE - 8;
+
+    public B2JsonBoundedByteArrayOutputStream() {
+        // initialize the outputBuffer to 64 bytes
+        this.output = new byte[INITIAL_CAPACITY];
+        this.size = 0;
+    }
+
+    /**
+     * writes one byte into the output array
+     *
+     * @param i one byte
+     * @throws IOException if new expanded capacity is over
+     *                     the MAX_CAPACITY
+     */
+    @Override
+    public synchronized void write(int i) throws IOException {
+        // expand capacity if necessary
+        if (size + 1 > output.length) {
+            final int newCapacity = expandCapacity(size + 1);
+
+            // allocate a new byte array with old content copied
+            output = Arrays.copyOf(output, newCapacity);
+        }
+
+        output[size] = (byte) i;
+        size++;
+    }
+
+    /**
+     * writes an array of bytes into the output array
+     *
+     * @param bytes input array of bytes
+     * @param offset offset for the input array
+     * @param length number of bytes to write
+     * @throws IOException if new expanded capacity is over
+     *                     the MAX_CAPACITY
+     */
+    @Override
+    public synchronized void write(byte[] bytes, int offset, int length) throws IOException {
+        // expand capacity if necessary
+        if (size + length > output.length) {
```

Comment:
maybe factor out `size + length` -- it's computed twice, and giving it a name would be nice